feat: documentation on using versioning (#226)
RonasIT/laravel-helpers: documentation/versioning.md

[<< Helpers][1]
[Traits >>][2]

# Versioning

## Route Macros
The macros provide a convenient way to manage API versioning in Laravel routes.

They allow you to specify which versions of the API a route is accessible for.

By using these macros, routes can be easily grouped and protected according to version-specific logic,
ensuring backward compatibility and smooth evolution of the API.

### versionRange
The versionRange macro allows you to define a range of API versions for which a route is accessible.
```
Route::versionRange(VersionEnum::v1, VersionEnum::v2)
```
### versionFrom

> This macro helps define the minimum API version required to access a route.

Reviewer: ```suggestion
Allows to define the minimum API version required to access routes.
```

```
Route::versionFrom(VersionEnum::v1)
```
### versionTo

> The versionTo macro allows you to define the maximum API version for which a route is accessible.

Reviewer: ```suggestion
Allows to define the maximum API version allowed for routes.
```

```
Route::versionTo(VersionEnum::v2)
```
### version

> The version macro allows you to define a specific API version for which a route is accessible.

Reviewer: ```suggestion
Allows to define the only allowed API version for routes.
```

```
Route::version(VersionEnum::v1)
```

## Middlewares

### CheckVersionMiddleware
This middleware verifies the requested API version.
If the version is blocked in the configuration, the application returns HTTP_UPGRADE_STATUS.

To manage blocked versions, configure the `disabled_api_versions` parameter in your application `/app/config/app.php`.

Example:
```
'disabled_api_versions' => [
    VersionEnum::v1->value,
    VersionEnum::v3->value,
] 
```

> ### ClearVersion

Reviewer: let's create another one middleware `VersioningMiddleware` which will call both middlewares and described only it in the readme

and checking greater-than-or-equal and less-than-or-equal conditions.

This class is typically used to implement version-based logic in controllers or middleware.

## Testing helpers

The `json` method of the `Testing\TestCase` class has been extended and now will add the version prefix to each test API call, if the API version is set.

### setAPIVersion

Helper using to set the API version as a class field. We suggest setting the actual API version in the `setUp` method of the application's `TestCase` class. In this case - all tests will work with the same actual API version.

When implementing tests for a particular API version, the `setAPIversion` method may be called in the `setUp` method of the specific test class.

### withoutAPIVersion

For some types of tests, the API version may not be used e.g., for webhooks or health status APIs. In these cases the `withoutAPIVersion` helper call will prevent the adding of a version prefix to each API call in tests.

Helper may be called either in a single test or in the `setUp` method of the test case.

## Usage

### Step 1.

Create an Enum structure to store the list of API versions.

> Created Enum must implement the `VersionEnumContract` interface and use `EnumTrait`.

Reviewer: let's add links to the contract and trait files

```
<?php

namespace App\Enum;

use RonasIT\Support\Contracts\VersionEnumContract;
use RonasIT\Support\Traits\EnumTrait;

enum VersionEnum: string implements VersionEnumContract
{
    use EnumTrait;

    case v1 = '1';
}
```

### Step 2.

Bind the `VersionEnumContract` to the created `VersionEnum` in any service provider.

```
    public function register(): void
    {
        $this->app->bind(VersionEnumContract::class, fn () => VersionEnum::class);
    }
```

### Step 3.

Wrap all routes to the group of common versions

```
Route::prefix('v{version}')
    ->middleware(VersioningMiddleware::class)
    ->group(function () {
        Route::versionFrom(VersionEnum::v1)->group(function () {
            Route::middleware('auth_group')->group(function () {
                Route::controller(AuthController::class)->group(function () {
                    Route::post('auth/logout', 'logout');
                });
            });
            Route::middleware(['guest_group'])->group(function () {
                Route::controller(AuthController::class)->group(function () {
                    Route::post('login', 'login');
                });
            });
        });
    });
```

> ### Implement backward compatibility

Reviewer: ```suggestion
### Implement backward compatibility

```

```
Route::version(VersionEnum::v0_1)->group(function () {
    Route::controller(TestControllerV0_1::class)->group(function () {
        Route::get('tests/{id}', 'get')->whereNumber('id');
    });
});

Route::prefix('v{version}')
    ->middleware(VersioningMiddleware::class)
    ->group(function () {
        Route::versionFrom(VersionEnum::v0_1)->group(function () {
            Route::controller(TestController::class)->group(function () {

>                 Route::versionFrom(VersionEnum::v0_2)->group(function () {
>                      Route::get('tests/{id}', 'get')->whereNumber('id');
>                 });

Reviewer: please check that this examplework

```suggestion
                Route::versionFrom(VersionEnum::v0_2)->get('tests/{id}', 'get')->whereNumber('id');
```

Author: @DenTray I checked. It is work well

When implementing tests for a particular API version, the `setAPIversion` method may be called in the `setUp` method of the specific test class.

### withoutAPIVersion

For some types of tests, the API version may not be used e.g., for webhooks or health status APIs. In these cases the `withoutAPIVersion` helper call will prevent the adding of a version prefix to each API call in tests.

Helper may be called either in a single test or in the `setUp` method of the test case.

## Usage

### Step 1.

Create an Enum structure to store the list of API versions.

Created Enum must implement the [VersionEnumContract][6] interface and use [EnumTrait][7].

```
<?php

namespace App\Enum;

use RonasIT\Support\Contracts\VersionEnumContract;
use RonasIT\Support\Traits\EnumTrait;

enum VersionEnum: string implements VersionEnumContract
{
    use EnumTrait;

    case v1 = '1';
}
```

### Step 2.

Bind the `VersionEnumContract` to the created `VersionEnum` in any service provider.

```

>     public function register(): void
>     {
>         $this->app->bind(VersionEnumContract::class, fn () => VersionEnum::class);
>     }

Reviewer: ```suggestion
public function register(): void
{
    $this->app->bind(VersionEnumContract::class, fn () => VersionEnum::class);
}
```
